### ISSUE TITLE
impl `Stream` for Generator?

### ISSUE THREAD

**Issue by jakobhellermann** [whatisaphone/genawaiter#1]
would it be possible to implement `futures::Stream` for `Generator`, kind of like [futures-async-stream](https://github.com/taiki-e/futures-async-stream/blob/master/src/stream.rs#L36)?

I'm hoping `genawaiter` can be an alternative on stable for that crate.

**Comment by whatisaphone** [comment]
Yes, that seems doable!

It would probably make sense to put it behind `#[cfg(feature = "futures")]` so people can avoid the extra dependency if they want.

My plan of attack would probably look something like this:

- extend the [advance](https://github.com/whatisaphone/genawaiter/blob/414dc5c99b404b510d71343e019cbf73faf816f3/src/rc/engine.rs#L20) functions to optionally accept a `Context` to use instead of creating one.
- add `impl<G: Generator<etc>> Stream for G`, which passes along the context to `advance`
- update the `advance` logic to handle the new data flow paths

Seems simple enough, but the devil is in the details, of course

**Comment by DevinR528** [comment]
I've gotten something sort of working as a [stream](https://github.com/DevinR528/genawaiter/blob/gen-stream/src/stack/generator.rs#L194) but because of this blanket impl in `futures_core::Stream`
```bash
error[E0119]: conflicting implementations of trait `futures_core::stream::Stream` for type `&mut _`:
   --> src/stack/generator.rs:172:1
    |
172 | impl<G: Generator> Stream for G {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `futures_core`:
            - impl<S> futures_core::stream::Stream for &mut S
              where S: futures_core::stream::Stream, S: std::marker::Unpin, S: ?Sized;
    = note: downstream crates may implement trait `ops::Coroutine` for type `&mut _`
```
I also ran into type resolving problems when trying to `impl Stream for Gen<'s, Y, (), F>`. Just to get it to work so i could investigate further I did
```rust
impl<'s, Y, F: Future> StreamGen<'s, Y, F> {
    pub fn new(generator: Gen<'s, Y, (), F>) -> Self {
        StreamGen { generator }
    }
}

impl<'s, Y, F: Future> Stream for StreamGen<'s, Y, F> {
    type Item = Y;

    fn poll_next(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Self::Item>> {
        let this = self.get_mut();
        match this.generator.resume_with_pass_ctx((), cx) {
            GeneratorState::Yielded(x) => {
                Poll::Ready(Some(x))
            }
            GeneratorState::Complete(_f) => {
                Poll::Ready(None)
            }
        }
    }
}
```
I was wondering how this would be used? Only a few of the `StreamExt` methods seem to work,
`take()`, `next()`,`chunks()` a few others but I threw this thing together to make sure it knew it was a stream at least.
```rust
let gen = &mut StreamGen::new(generator);
    let cmp = &[1_i32, 3, 5, 7, 9];
    futures::executor::block_on(async {
        for x in cmp.iter() {
            let y = gen.next().await.unwrap();
            assert_eq!(*x, y)
        }
    });
```

**Comment by whatisaphone** [comment]
@DevinR528 Thanks for taking a look at this!

It looks like you ran into the infamous [coherence rules](https://github.com/Ixrec/rust-orphan-rules). If you have these impls:

```rust
impl<G: Generator> Stream for G {}
impl<S: Stream>    Stream for &mut S {}
```

If a type is both a `&mut _` and a `Generator`, it would be ambiguous which `impl` applies.

I took a look at your branch, and (correct me if I'm wrong) it seems like it basically just replicates [`futures::stream::iter`](https://docs.rs/futures/0.3.1/futures/stream/fn.iter.html), which converts an iterator to a stream. If that's what this issue is about, I don't think it needs to be a part of this crate, since it's already part of `futures`.

Here's some code with `Stream`s that you can write today, using the genawaiter on crates.io:

```rust
use futures::executor::block_on_stream;
use genawaiter::sync::Gen;

#[test]
fn test() {
    let gen = Gen::new(|co| async move {
        co.yield_(10).await;
        co.yield_(20).await;
    });

    let stream = futures::stream::iter(gen);
    let items: Vec<_> = block_on_stream(stream).collect();
    assert_eq!(items, [10, 20]);
}
```

To those watching this issue, does `futures::stream::iter` work for your use case?

**Comment by DevinR528** [comment]
Yeah I agree, I was mostly interested in learning more about the new `std::future` and `async/await` which I did thanks to this crate!! I don't have a use case so `futures::stream::iter` sounds good to me.

**Comment by jjakobh** [comment]
my use case was basically the following:
```rust
loop {
    match socket.recv(&mut buf).await {
        Err(e) if e.kind() == TimedOut => break,
        Ok(v) => yield v;
    }
}
```
I tried rewriting it using `co.yield_(...)` and `futures::stream::iter` but then I run into the panic `A generator was awaited without first yielding a value. Inside a generator, do not await any futures other than the one returned by 'Co::yield_'.`

How I can await a Future while using a `Gen`?

**Comment by whatisaphone** [comment]
Thanks for the snippet. I sketched out a design that I think would make that work and pushed a wip branch. [This test](https://github.com/whatisaphone/genawaiter/blob/8d4ee5d2ed950a2f9c10e93d1f05a4aaa53e6a31/src/sync/stream.rs#L48) (high level, `Stream`) and [this test](https://github.com/whatisaphone/genawaiter/blob/8d4ee5d2ed950a2f9c10e93d1f05a4aaa53e6a31/src/sync/mod.rs#L279) (lower-level `resume_async`) show how to use it. If you have time, it would be helpful if you tried out the branch and let me know if it does what you're after

**Comment by jjakobh** [comment]
thanks, it works just as expected!

**Comment by whatisaphone** [comment]
Released in v0.2.2! Make sure to add the new feature in `Cargo.toml`:

```toml
[dependencies]
genawaiter = { version = "0.2.2", features = ["futures03"] }
```
